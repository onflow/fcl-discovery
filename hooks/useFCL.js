import { useEffect, useState } from 'react'
import { WalletUtils } from '@onflow/fcl'

export function useFCL() {
  const [hasInitialized, setHasInitialized] = useState(false)
  const [loading, setLoading] = useState(false)
  const [appConfig, setAppConfig] = useState()
  const [clientConfig, setClientConfig] = useState()
  const [appVersion, setAppVersion] = useState(null)
  const [extensions, setExtensions] = useState([])
  const [walletInclude, setWalletInclude] = useState([])
  const [wcProjectId, setWcProjectId] = useState(null)
  const [wcPairings, setWcPairings] = useState([])

  useEffect(() => {
    setHasInitialized(true)
    setLoading(true)

    WalletUtils.ready(({ fclVersion, body, config }) => {
      // config.client.fclVersion is only available starting in version 0.0.79-alpha.4
      // config?.client?.extensions starts in fcl v1
      const appFclVersion = config?.client?.fclVersion || fclVersion || null
      const clientExtensions =
        config?.client?.extensions || body?.extensions || []
      const wc = config?.client?.wc || null

      if (config?.app) {
        setAppConfig(config.app)
      }

      if (config?.client) {
        setClientConfig(config.client)
      }

      if (appFclVersion) {
        setExtensions(clientExtensions)
        setAppVersion(appFclVersion)
        setWalletInclude(config.discoveryAuthnInclude || [])
      }

      if (wc?.projectId) {
        setWcProjectId(wc.projectId)
        setWcPairings(wc.pairings)
      }

      setLoading(false)
    })
  }, [])

  return {
    hasInitialized,
    loading,
    appConfig,
    clientConfig,
    appVersion,
    extensions,
    walletInclude,
    wcProjectId,
    wcPairings,
  }
}
