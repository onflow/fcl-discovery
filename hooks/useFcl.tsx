import { useEffect, useRef, useState } from 'react'
import { WalletUtils } from '@onflow/fcl'
import { RpcClient } from '@onflow/util-rpc'
import { Service, Strategy } from '../types'
import { CUSTOM_RPC } from '../helpers/constants'
import {
  DiscoveryNotification,
  FclRequests,
  FclRpcClient,
} from '../helpers/rpc'

type WalletUtilsProps = {
  fclVersion: string
  body: { [key: string]: any }
  config: { [key: string]: any }
}

export interface FclConfig {
  appConfig: { [key: string]: any }
  clientConfig: { [key: string]: any }
  appVersion: string
  walletInclude: string[]
  clientServices: Service[]
  supportedStrategies: Strategy[]
  rpcEnabled?: boolean
}

export function useFcl() {
  const [config, setConfig] = useState<FclConfig | null>(null)
  const [error, setError] = useState<string | null>(null)
  const timeout = useRef<NodeJS.Timeout | null>(null)

  const [rpc, setRpc] = useState<FclRpcClient | null>(null)
  const rpcEnabled = config?.rpcEnabled

  useEffect(() => {
    try {
      WalletUtils.ready(({ fclVersion, body, config }: WalletUtilsProps) => {
        // config.client.fclVersion is only available starting in version 0.0.79-alpha.4
        // config?.client?.extensions starts in fcl v1
        const state = {
          appConfig: config.app,
          clientConfig: config.client,
          appVersion: config.client?.fclVersion || fclVersion || null,
          walletInclude:
            config.discoveryAuthnInclude ||
            config.client.discoveryAuthnInclude ||
            [],
          clientServices:
            config.client?.clientServices ||
            config.client?.extensions ||
            body.extensions ||
            [],
          supportedStrategies: config.client?.supportedStrategies || [],
          rpcEnabled: config.client?.discoveryRpcEnabled || false,
        } as FclConfig

        setConfig(state)
        clearTimeout(timeout.current)
      })

      timeout.current = setTimeout(() => {
        setError(
          'Error occured, if you are the developer please check the console for more information.',
        )
        console.error(
          'Timeout: No response from FCL received within 5s - please ensure the application you are trying to connect to is running FCL & has the correct configuration and try again.',
        )
      }, 5000)
    } catch (e) {
      clearTimeout(timeout.current)
      setError(
        'Error occured, if you are the developer please check the console for more information.',
      )

      console.error(e)
      console.error(
        "An error has occured connecting to the dApp's FCL instance, please see docs: https://developers.flow.com/tools/fcl-js/reference/discovery",
      )
    }

    return () => {
      clearTimeout(timeout.current)
    }
  }, [])

  useEffect(() => {
    if (!rpcEnabled || rpc) return

    const _rpc = new RpcClient<FclRequests, {}>({
      notifications: Object.values(DiscoveryNotification),
    })

    // Bind receiver for messages from FCL
    WalletUtils.onMessageFromFCL(
      CUSTOM_RPC,
      ({ payload: msg }: { payload: any }) => {
        _rpc.receive(msg)
      },
    )

    // Bind sender for messages to FCL
    _rpc.connect({
      send: msg => {
        WalletUtils.sendMsgToFCL(CUSTOM_RPC, { payload: msg })
      },
    })

    setRpc(_rpc)
  }, [rpcEnabled, rpc])

  return {
    error,
    config,
    rpc,
    isLoading: !config && !error,
  }
}
