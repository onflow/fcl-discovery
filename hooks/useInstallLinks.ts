import { useCallback, useMemo } from 'react'
import { Wallet } from '../data/wallets'
import { useDevice } from '../contexts/DeviceContext'
import { DeviceType } from '../helpers/device-info'
import { useConfig } from '../contexts/FclContext'
import { FCL_SERVICE_METHODS } from '../helpers/constants'

export function useInstallLinks(wallet: Wallet) {
  const { deviceInfo } = useDevice()
  const { supportedStrategies } = useConfig()

  const getInstallLink = useCallback(
    (strategy: FCL_SERVICE_METHODS) => {
      let installLink: string | null = null
      switch (strategy) {
        case FCL_SERVICE_METHODS.WC:
          if (deviceInfo.type === DeviceType.MOBILE) {
            installLink =
              wallet.installLink?.[deviceInfo.platform] ||
              wallet.installLink?.mobile
          } else {
            installLink = wallet.installLink?.mobile
          }
          break
        case FCL_SERVICE_METHODS.EXT:
          if (deviceInfo.type !== DeviceType.DESKTOP) {
            return null
          }

          installLink =
            wallet.installLink?.[deviceInfo.browser] ||
            wallet.installLink?.browserExtension

          if (!installLink) {
            return null
          }
          break
      }

      return installLink
    },
    [deviceInfo, wallet],
  )

  const links = useMemo(() => {
    if (!wallet) {
      return {}
    }

    const res = supportedStrategies.reduce(
      (acc, strategy) => {
        let installLink: string | null = getInstallLink(strategy)

        if (!installLink) {
          return acc
        }
        return { ...acc, [strategy]: installLink }
      },
      {} as Record<FCL_SERVICE_METHODS, string>,
    )
    return res
  }, [deviceInfo, wallet, supportedStrategies])

  return links
}
