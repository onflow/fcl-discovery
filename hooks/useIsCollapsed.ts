import { useBreakpointValue } from '@chakra-ui/react'
import { useDevice } from '../contexts/DeviceContext'
import { DeviceType } from '../helpers/device'

export function useIsCollapsed() {
  const { deviceInfo } = useDevice()
  const isCollapsed = useBreakpointValue(
    { base: true, md: false },
    {
      ssr: true,
      fallback:
        deviceInfo.type === DeviceType.MOBILE && !deviceInfo.isTablet
          ? 'md'
          : 'base',
    },
  )
  return isCollapsed
}
