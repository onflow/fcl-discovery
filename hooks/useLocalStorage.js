import { useState, useEffect } from 'react'

function getStorageValue(key, defaultValue) {
  const saved = localStorage.getItem(key)
  const initial = JSON.parse(saved)
  return initial || defaultValue
}

export default function useLocalStorage(key, defaultValue) {
  const [value, setValue] = useState(() => getStorageValue(key, defaultValue))

  useEffect(() => {
    localStorage.setItem(key, JSON.stringify(value))
  }, [key, value])

  return [value, setValue]
}
