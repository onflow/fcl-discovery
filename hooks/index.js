import useFCL from './useFCL'
import useLocalStorage from './useLocalStorage'

export { useFCL, useLocalStorage }
