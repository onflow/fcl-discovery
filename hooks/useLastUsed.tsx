import { LOCAL_STORAGE_KEYS } from '../helpers/constants'
import { useLocalStorage } from './useLocalStorage'

export function useLastUsed() {
  const [, setLastUsed] = useLocalStorage(LOCAL_STORAGE_KEYS.LAST_USED, null)
  const [lastUsed] = useLocalStorage(LOCAL_STORAGE_KEYS.LAST_USED, null)

  return { lastUsed, setLastUsed }
}
