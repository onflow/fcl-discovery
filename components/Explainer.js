import styled from "styled-components"
import {COLORS} from "../helpers/constants"
import {useFCL} from "../hooks/useFCL"

const ExplainerContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 30px 0;
`

const ExplainerDescription = styled.div`
  font-size: 1.2rem;
  margin-bottom: 5px;
`

const ExplainerLink = styled.div`
  text-decoration: none;
  color: ${COLORS.SECONDARY};
  cursor: pointer;
`

export default function Explainer() {
  const {appConfig} = useFCL()

  return (
    <ExplainerContainer>
      <ExplainerDescription>
        <strong>{appConfig?.title || "Unknown App"}</strong> wants to connect to your wallet.
      </ExplainerDescription>
      <ExplainerLink>
        <div onClick={() => window.open("https://docs.onflow.org/flow-token/available-wallets/", "_blank")}>What is a wallet?</div>
      </ExplainerLink>
    </ExplainerContainer>
  )
}
