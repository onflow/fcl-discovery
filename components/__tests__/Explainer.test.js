import { render } from '@testing-library/react'
import Explainer from '../Explainer'
import * as hooks from '../../hooks/useFCL'

describe('Component: Explainer', () => {
  let fclHookSpy

  beforeEach(() => {
    fclHookSpy = jest.spyOn(hooks, 'useFCL')
  })

  afterEach(() => {
    fclHookSpy.mockRestore()
  })

  test('should render the component with app title', () => {
    fclHookSpy.mockImplementation(() => ({
      appConfig: {
        title: 'Test App',
      },
    }))

    const { container } = render(<Explainer />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('should render app as unknown in Explainer component if no title', () => {
    fclHookSpy.mockImplementation(() => ({
      appConfig: {
        title: null,
      },
    }))

    const { container } = render(<Explainer />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
