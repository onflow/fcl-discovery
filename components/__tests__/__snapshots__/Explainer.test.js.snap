// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Explainer should render app as unknown in Explainer component if no title 1`] = `
<div
  class="sc-bczRLJ dhqKmX"
>
  <div
    class="sc-gsnTZi gdkTFb"
  >
    <div>
      <strong>
        Unknown App
      </strong>
    </div>
    <div>
      wants to connect to your wallet.
    </div>
  </div>
</div>
`;

exports[`Component: Explainer should render the component with app title 1`] = `
<div
  class="sc-bczRLJ dhqKmX"
>
  <div
    class="sc-gsnTZi gdkTFb"
  >
    <div>
      <strong>
        Test App
      </strong>
    </div>
    <div>
      wants to connect to your wallet.
    </div>
  </div>
</div>
`;
